### PR TITLE
DEP: `stats.ttest_ind`: deprecate `random_state` and `permutation` in favor of `method`; require keywords for most args

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-21833

#### What does this implement/fix?
The `permutation` and `random_state` arguments of `ttest_ind` were added in gh-4824 to enable the user to perform a permutation version of the t-test. A standard API for performing resampling versions of tests using a `method` argument was developed in gh-18227. The SPEC 7 deprecation of the `random_state` arguments provides some impetus for transitioning from `permutation` and `random_state` to `method`. 

This PR makes the following changes to `ttest_ind`:
- adds a `method` argument that allows the user to configure permutation and Monte Carlo versions of the test
- deprecates `permutation` and `random_state` arguments in favor of `method`
- deprecates positional use of all arguments of `ttest_ind` after `x` and `y`. The removal of `permutation` and `random_state` would *also* require later arguments (`alternative` and `trim`) to become keyword-only, anyway; we might as well make all the optional parameters keyword-only at the same time.

The transition to the new interface is quite simple:
```python3
from scipy import stats
... # define variables

# This:
stats.ttest_ind(a, b, 0, True, 'propagate', 9999, rng, 'two-sided', 0)

# becomes
method = stats.PermutationMethod(n_resamples=9999, rng=rng)
stats.ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate',
                method=method, trim=0)
```

### CODE REVIEW

File: scipy/_lib/deprecation.py [hunk 1]
```diff
@@ -206,14 +214,26 @@ def _inner_deprecate_positional_args(f):
...
+
             extra_args = len(args) - len(all_args)
             if extra_args <= 0:
+                warn_deprecated_args(kwargs)
                 return f(*args, **kwargs)
 
             # extra_args > 0
-            args_msg = ", ".join(kwonly_args[:extra_args])
+            kwonly_extra_args = set(kwonly_args[:extra_args]) - deprecated_args
+            args_msg = ", ".join(kwonly_extra_args)
```
Comment (mdhaber) [inline scipy/_lib/deprecation.py:236]: This warns about the deprecated *positional* use of arguments passed by position. We don't want the `deprecated_args` (which are completely decorated) to be included in that list; warn about those separately.

File: scipy/_lib/deprecation.py [hunk 2]
```diff
@@ -225,8 +245,25 @@ def inner_f(*args, **kwargs):
...
             kwargs.update(zip(sig.parameters, args))
+            warn_deprecated_args(kwargs)
             return f(**kwargs)
 
+        doc = FunctionDoc(inner_f)
+        kwonly_extra_args = set(kwonly_args) - deprecated_args
+        admonition = f"""
+.. deprecated:: {version}
+    Use of argument(s) ``{kwonly_extra_args}`` by position is deprecated; beginning in 
+    SciPy {version}, these will be keyword-only. """
```
Comment (mdhaber) [inline scipy/_lib/deprecation.py:256]: Ugly indentation; I know, but it makes it easy to keep Sphinx happy.

File: scipy/stats/_stats_py.py [hunk 3]
```diff
@@ -6334,6 +6352,11 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate',
 
     By default, the p-value is determined by comparing the t-statistic of the
     observed data against a theoretical t-distribution.
+
+    (In the following, note that the argument `permutations` itself is
+    deprecated, but a nearly identical test may be performed by creating
+    an instance of `scipy.stats.PermutationMethod` with ``n_resamples=permutuations``
+    and passing it as the `method` argument.)
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:6359]: We could just remove all of this temporarily. We will be writing an interactive notebook tutorial for permutation and Monte Carlo versions of the t-test (like https://scipy.github.io/devdocs/tutorial/stats/hypothesis_normaltest.html) shortly, and we will link to that from here.

File: scipy/stats/tests/test_resampling.py [hunk 4]
```diff
@@ -1523,35 +1523,6 @@ def statistic1d(x):
         assert res.statistic == res2.statistic
         assert_allclose(res.pvalue, res2.pvalue, atol=1e-2)
 
-    @pytest.mark.parametrize('alternative', ('less', 'greater'))
-    # Different conventions for two-sided p-value here VS ttest_ind.
-    # Eventually, we can add multiple options for the two-sided alternative
-    # here in permutation_test.
-    @pytest.mark.parametrize('permutations', (30, 1e9))
-    @pytest.mark.parametrize('axis', (0, 1, 2))
-    def test_against_permutation_ttest(self, alternative, permutations, axis):
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:1532]: This tested `scipy.stats.permutation_test` against using the `permutations` argument of `ttest_ind`, but there are other, stronger tests now.

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -5343,6 +5343,7 @@ def test_ttest_ind_scalar():
     assert np.isnan(p)
 
 
+@pytest.mark.filterwarnings("ignore:Arguments...:DeprecationWarning")
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5346]: We test for the warning about deprecated `permutations` below.

I suppose there are no tests for deprecated positional use, but I assume the `_deprecate_positional_args` decorator was already tested, so I'm just testing the new part.

File: scipy/stats/tests/test_stats.py [hunk 6]
```diff
@@ -5589,6 +5590,65 @@ def test_ttest_ind_permutation_check_p_values(self):
...
+        assert_equal(res.pvalue, ref.pvalue)
+
+        # Sanity check against theoretical t-test
+        ref = stats.ttest_ind(x, y, axis=-1, alternative=alternative)
+        assert_equal(res.statistic, ref.statistic)
+        assert_allclose(res.pvalue, ref.pvalue, rtol=3e-2)
+
+    @pytest.mark.parametrize("alternative", ['less', 'greater', 'two-sided'])
+    @pytest.mark.parametrize("shape", [(12,), (2, 12)])
+    def test_monte_carlo_method(self, alternative, shape):
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5623]: I tried combining this with the previous test, but this is really more readable.

We could also consider removing the Monte Carlo version of the test entirely during this (complicated) transition period, just to keep things a little simpler.

**Review by mdhaber — COMMENTED** [review]
Some comments to aid the reviewer.

**Comment by j-bowhay** [comment]
Could you write a post to the forum regarding the deprecation?

File: scipy/stats/_stats_py.py [hunk 7]
```diff
@@ -6479,6 +6492,15 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate',
     if not (0 <= trim < .5):
         raise ValueError("Trimming percentage should be 0 <= `trim` < .5.")
 
+    if not isinstance(method, PermutationMethod | MonteCarloMethod | None):
+        message = ("`method` must be an instance of `PermutationMethod`, an instance "
+                   "of `MonteCarloMethod`, or None (default).")
+        raise ValueError(message)
```
Comment (j-bowhay) [inline scipy/stats/_stats_py.py:6498]: Can we add a test for this

File: scipy/stats/_stats_py.py [hunk 8]
```diff
@@ -6479,6 +6492,15 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate',
...
+
+    if not is_numpy(xp) and method is not None:
+        message = "Use of resampling methods is compatible only with NumPy arrays."
+        raise NotImplementedError(message)
```
Comment (j-bowhay) [inline scipy/stats/_stats_py.py:6502]: And this

File: scipy/stats/tests/test_stats.py [hunk 9]
```diff
@@ -5797,7 +5858,9 @@ def test_errors_unsupported(self):
         # confirm that attempting to trim with permutations raises an error
         match = "Use of `permutations` is incompatible with with use of `trim`."
         with assert_raises(NotImplementedError, match=match):
-            stats.ttest_ind([1, 2], [2, 3], trim=.2, permutations=2)
+            message = "Arguments {'permutations'} are deprecated, whether..."
+            with pytest.warns(DeprecationWarning, match=message):
+                stats.ttest_ind([1, 2], [2, 3], trim=.2, permutations=2)
```
Comment (j-bowhay) [inline scipy/stats/tests/test_stats.py:5863]: Can we test randomatate also raises a warning

**Review by j-bowhay — COMMENTED** [review]
A few comments about tests but otherwise looks good

File: scipy/stats/tests/test_stats.py [hunk 10]
```diff
@@ -5589,20 +5590,95 @@ def test_ttest_ind_permutation_check_p_values(self):
...
+        assert_equal(res.statistic, ref.statistic)
+        assert_equal(res.pvalue, ref.pvalue)
+
+        # Sanity check against theoretical t-test
+        ref = stats.ttest_ind(x, y, axis=-1, alternative=alternative)
+        assert_equal(res.statistic, ref.statistic)
+        assert_allclose(res.pvalue, ref.pvalue, rtol=6e-2)
+
+    def test_resampling_input_validation(self):
+        rng = np.random.default_rng(328924823)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:5652]: Will commit with `[lint only]` if the rest looks good.
```suggestion
```

**Review by j-bowhay — APPROVED** [review]
Thanks! Can you make sure this makes it into the release notes

**Pull request merged**